src/wire/igmpv2.rs

```diff
@@ -0,0 +1,417 @@
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Return the message type field.
+    #[inline]
+    pub fn msg_type(&self) -> Message {
+        let data = self.buffer.as_ref();
+        Message::from(data[field::TYPE])
+    }
+
+    /// Return the Max reponse time
+    ///
+    /// See [RFC 3376]: 4.1.1. Max Resp Code
+    ///
+    /// [RFC 3376]: https://tools.ietf.org/html/rfc3376
+    #[inline]
+    pub fn max_resp_time(&self) -> u16 {
+        let data = self.buffer.as_ref();
+        let value = data[field::MAX_RESP_TIME].into();
+        if value < 128 {
+            value
+        } else {
+            let mant = value & 0xF;
+            let exp = (value >> 4) & 0x7;
+            (mant | 0x10) << (exp + 3)
+        }
+    }
+
+    /// Return the checksum field.
+    #[inline]
+    pub fn checksum(&self) -> u16 {
+        let data = self.buffer.as_ref();
+        NetworkEndian::read_u16(&data[field::CHECKSUM])
+    }
+
+    /// Return the source address field.
+    #[inline]
+    pub fn group_addr(&self) -> Ipv4Address {
+        let data = self.buffer.as_ref();
+        Ipv4Address::from_bytes(&data[field::GROUP_ADDRESS])
+    }
+
+    /// Validate the header checksum.
+    ///
+    /// # Fuzzing
+    /// This function always returns `true` when fuzzing.
+    pub fn verify_checksum(&self) -> bool {
+        if cfg!(fuzzing) {
+            return true;
+        }
+
+        let data = self.buffer.as_ref();
+        checksum::data(data) == !0
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> Packet<T> {
+    /// Set the message type field.
+    #[inline]
+    pub fn set_msg_type(&mut self, value: Message) {
+        let data = self.buffer.as_mut();
+        data[field::TYPE] = value.into()
+    }
+
+    /// Set the Max Resp Time field.
+    #[inline]
+    pub fn set_max_resp_time(&mut self, value: u8) {
+        let data = self.buffer.as_mut();
+        data[field::MAX_RESP_TIME] = value
+    }
+
+    /// Set the checksum field.
+    #[inline]
+    pub fn set_checksum(&mut self, value: u16) {
+        let data = self.buffer.as_mut();
+        NetworkEndian::write_u16(&mut data[field::CHECKSUM], value)
+    }
+
+    /// Set the group address field
+    #[inline]
+    pub fn set_group_address(&mut self, addr: Ipv4Address) {
+        let data = self.buffer.as_mut();
+        data[field::GROUP_ADDRESS].copy_from_slice(addr.as_bytes());
+    }
+
+    /// Compute and fill in the header checksum.
+    pub fn fill_checksum(&mut self) {
+        self.set_checksum(0);
+        let checksum = {
+            let data = self.buffer.as_ref();
+            !checksum::data(data)
+        };
+        self.set_checksum(checksum)
+    }
+}
+
+
+/// A high-level representation of an Internet Group Management Protocol v2 header.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Repr {
+    MembershipQuery {
+        /// Maximum Response Time in 1/10th second (100ms) units
+        max_resp_time: u16,
```

Comment:
Since not all `u16` values are valid for `max_resp_time` anyway, I think you should use `Duration` here, and silently clamp values to the closest valid ones.